### PR TITLE
Issue #6 Create settings classes and add quiz settings to rule.php

### PR DESCRIPTION
**Pull request by andrewmadden**

### CODE REVIEW

**Comment by andrewmadden** [comment]
Still pretty rough - Some feedback about structure would be helpful.

File: classes/config_settings.php [hunk 1]
```diff
@@ -0,0 +1,259 @@
...
+    /**
+     * Static constructor to create settings from existing quiz.
+     *
+     * @param string $quizid ID of quiz that the settings belong to.
+     * @return config_settings This instance.
+     *
+     * @throws \dml_exception
+     * @throws \moodle_exception
+     */
+    public static function with_quizid(string $quizid) {
```
Comment (dmitriim) [inline classes/config_settings.php]: As per  https://docs.moodle.org/dev/Coding_style#Functions_and_Methods "Verbosity is encouraged: function names should be as illustrative as is practical to enhance understanding." Would suggest to pick a new name.

File: classes/config_settings.php [hunk 2]
```diff
@@ -0,0 +1,259 @@
...
+    }
+
+    /**
+     * Check if the settings exist.
+     *
+     * @return bool Whether any new or existing settings were found for quiz.
+     */
+    public function exists() {
+        // Cast settings to array as empty stdClass will return true.
+        return !empty((array) $this->settings);
```
Comment (dmitriim) [inline classes/config_settings.php]: What would happen if I create a class with bunch of random settings? Do we need to hardcore settings we support (e.g. settings we  have fields for in DB)?

File: classes/config_settings.php [hunk 3]
```diff
@@ -0,0 +1,259 @@
...
+                $this->existingsettingsid = $DB->insert_record(self::SETTINGS_TABLE, $this->settings);
+                // Check if the insert was successful, and if not, reset existingsettingsid to -1.
+                if ($this->existingsettingsid === false) {
+                    $this->existingsettingsid = -1;
+                } else {
+                    $success = true;
+                }
+            }
+        } catch (\dml_exception $e) {
+            debugging('quizaccess_seb - save settings: ' . $e->getMessage());
```
Comment (dmitriim) [inline classes/config_settings.php]: I think we should throw an exception here and not suppress it

File: classes/config_settings.php [hunk 4]
```diff
@@ -0,0 +1,259 @@
...
+
+    /**
+     * Delete files and settings from DB for this quiz.
+     *
+     * @return bool Whether or not settings were deleted successfully.
+     *
+     * @throws \coding_exception
+     * @throws \dml_exception
+     */
+    public function delete_settings() : bool {
```
Comment (dmitriim) [inline classes/config_settings.php]: looking at the names of methods... as long as we are in the class already, it feels like we can drop "_settings" postfix/ E.g. delete_settings() -> delete()

File: classes/config_settings.php (hunk 2 above)
Comment (dmitriim) [reply classes/config_settings.php]: Ah I see you have get_required_settings_keys! :+1:

File: rule.php [hunk 5]
```diff
@@ -62,6 +65,39 @@ public static function make (quiz $quizobj, $timenow, $canignoretimelimits) {
...
+        $elements = $settingsmanager->get_quiz_element_types();
+        $defaults = $settingsmanager->get_quiz_defaults();
+        $hideifs = $settingsmanager->get_quiz_hideifs();
+
+        // Get existing form data.
+        $quizid = $quizform->get_coursemodule()->instance;
+        $existingsettings = $DB->get_record('quizaccess_seb_quizsettings', ['quizid' => $quizid], '*');
+
+        // Insert all the form elements before the 'security' section as a group.
+        foreach ($elements as $name => $type) {
```
Comment (dmitriim) [inline rule.php]: it feels like we can just $settingsmanager->get_quiz_element_types() here and remove $elements

File: rule.php [hunk 6]
```diff
@@ -86,6 +122,18 @@ public static function validate_settings_form_fields(array $errors,
      *      which is the id of the quiz being saved.
      */
     public static function save_settings($quiz) {
+        $settingsmanager = new settings_manager();
```
Comment (dmitriim) [inline rule.php]: can we abstract this please and have something like self::get_setting_manager()

File: classes/config_settings.php [hunk 7]
```diff
@@ -0,0 +1,259 @@
...
+            $existingsettings = new \stdClass();
+        } else {
+            $this->existingsettingsid = $existingsettings->id;
+        }
+
+        // If new settings are empty, set existing settings, otherwise merge existing settings with new settings.
+        if (empty((array) $settings)) {
+            $this->settings = $existingsettings;
+        } else {
+            $this->settings = $this->merge_new_and_existing_settings($this->strip_prefix($settings), $existingsettings);
```
Comment (dmitriim) [inline classes/config_settings.php]: I can't see that you validate incoming $settings. Maybe it's worth to spend some time and look at https://docs.moodle.org/dev/Persistent and rewrite the code? it seems like it has out of the box validation.

Comment (dmitriim) [reply classes/config_settings.php]: or should __construct be private and we build the class only with "with_quizid" and "with_form_data" (where data is validated by the form) ?

File: classes/config_settings.php [hunk 8]
```diff
@@ -0,0 +1,259 @@
...
+    public function get_settings() : \stdClass {
+        return $this->settings;
+    }
+
+    /**
+     * Get the itemid of the upoaded seb config file if it exists.
+     *
+     * @return string Item ID of stored file.
+     */
+    public function get_config_itemid() {
```
Comment (dmitriim) [inline classes/config_settings.php]: let's remove this code for now

File: rule.php [hunk 9]
```diff
@@ -62,6 +65,39 @@ public static function make (quiz $quizobj, $timenow, $canignoretimelimits) {
      * @param MoodleQuickForm $mform the wrapped MoodleQuickForm.
      */
     public static function add_settings_form_fields(mod_quiz_mod_form $quizform, MoodleQuickForm $mform) {
+        global $DB, $USER;
+        $settingsmanager = new settings_manager();
```
Comment (dmitriim) [inline rule.php]: Love how you abstract all fields in the manager. :100:  Made usually messy form code more readable

File: classes/settings_manager.php [hunk 10]
```diff
@@ -0,0 +1,217 @@
...
+ * @author     Andrew Madden <andrewmadden@catalyst-au.net>
+ * @copyright  2019 Catalyst IT
+ * @license    http://www.gnu.org/copyleft/gpl.html GNU GPL v3 or later
+ */
+
+namespace quizaccess_seb;
+
+defined('MOODLE_INTERNAL') || die();
+
+class settings_manager {
```
Comment (dmitriim) [inline classes/settings_manager.php]: Is there any reason why it's 'manager'? will it have actually manager functionality?

File: classes/settings_manager.php [hunk 11]
```diff
@@ -0,0 +1,217 @@
...
+
+    /**
+     * Get all form elements that are a 'filepicker' type.
+     *
+     * @param bool $strippedprefix Defaults to true.
+     * @return array
+     */
+    public function get_file_picker_elements($strippedprefix = true) : array {
```
Comment (dmitriim) [inline classes/settings_manager.php]: Lat's remove this code as it's not getting used.

File: classes/settings_manager.php [hunk 12]
```diff
@@ -0,0 +1,217 @@
...
+        $formelements = $this->get_quiz_element_types();
+        $filepickers = [];
+        foreach ($formelements as $name => $type) {
+            if ($type == 'filepicker') {
+                if ($strippedprefix) {
+                    // Strip key prefix.
+                    $name = preg_replace("/^seb_/", "", $name);
```
Comment (dmitriim) [inline classes/settings_manager.php]: I think we do that in other place, maybe abstract to a method?

File: rule.php [hunk 13]
```diff
@@ -86,6 +122,18 @@ public static function validate_settings_form_fields(array $errors,
      *      which is the id of the quiz being saved.
      */
     public static function save_settings($quiz) {
+        $settingsmanager = new settings_manager();
+
+        $defaults = $settingsmanager->get_quiz_defaults();
+        // Set each required setting to default it not provided.
```
Comment (dmitriim) [inline rule.php]: typo? " it not provided" -> " if not provided"

File: rule.php [hunk 14]
```diff
@@ -86,6 +122,18 @@ public static function validate_settings_form_fields(array $errors,
...
+        foreach ($defaults as $name => $defaultvalue) {
+            if (!isset($quiz->$name)) {
+                $quiz->$name = $defaultvalue;
+            }
+        }
+
+        // Save/update settings from form.
+        config_settings::with_form_data($quiz)->save_settings();
```
Comment (dmitriim) [inline rule.php]: Can't see where we validate incoming text data. E.g. "seb_expressionsblocked" or "seb_expressionsallowed". Should we validate them? If so then create a card for that. We will treat it as a feature and do later.

File: classes/config_settings.php [hunk 15]
```diff
@@ -0,0 +1,259 @@
...
+     *
+     * @param int $quizid ID of quiz that the settings belong to.
+     * @param \stdClass|null $settings New quiz settings. May be provided by quiz settings form.
+     * @throws \dml_exception
+     * @throws \moodle_exception
+     */
+    public function __construct(int $quizid, \stdClass $settings = null) {
+        global $DB;
+
+        $this->quizid = $quizid;
```
Comment (dmitriim) [inline classes/config_settings.php]: Should we check if quiz with that ID exists at all?

File: classes/config_settings.php [hunk 16]
```diff
@@ -0,0 +1,259 @@
...
+        $existingsettings = $DB->get_record(self::SETTINGS_TABLE, ['quizid' => $quizid], '*');
+        if (empty($existingsettings)) {
+            $existingsettings = new \stdClass();
+        } else {
+            $this->existingsettingsid = $existingsettings->id;
+        }
+
+        // If new settings are empty, set existing settings, otherwise merge existing settings with new settings.
+        if (empty((array) $settings)) {
+            $this->settings = $existingsettings;
```
Comment (dmitriim) [inline classes/config_settings.php]: As settings could be potentially empty. Do we want to update DB definition and have default values of required fields?

Comment (dmitriim) [reply classes/config_settings.php]: Ah I can see that you validate it later. But still doesn't check if it matches DB type.

**Comment by andrewmadden** [comment]
* Now using a persistent class to represent the quiz settings. 
* Updated rule to use the persistent object.
* Renamed the `settings_manager` to `settings_provider` and stripped it down so it just provides the quiz settings info.
* Added tests for rule.php.
* Removed tests for `config_settings` and `settings_manager` because they are covered by persistent now.

**Pull request merged**